### ISSUE TITLE
Zombie memory leak in sites with iFrames

### ISSUE THREAD

**Issue by yaronassa** [assaf/zombie#995]
## The issue
**TL;DR**: Zombie doesn't clear timers for inner frames, and the pipeline doesn't wait for all requests to drain. This keeps orphaned objects even when destroying the browser instance.
I've finally managed to reproduce the issues we've been experiencing in our testing env. with an external site.

**In detail**:
I've built an example with a `cleanMemory` variable. Setting it to true will perform the manual cleanup Zombie should be doing - clearing timers in inner frames, and waiting for the pipeline to finish processing all the requests.
I ran this example with `--expose-gc` switch to make sure I can force the gc.

## The example:
```javascript
var Promise = require('bluebird');

var Browser;
var browser;

var visits = 0;
var maxVisits = 10;
var testPageURL = 'http://sampleinterviewquestion.com/'; //Just a site I've found the reproduces the leak we get in out inner testing environment.

var pendingRequests = []; //Store requests for manual cleanup

var cleanMemory = true; //Change this to switch manual cleanup on / off

//Actually browse for [maxVisits] cycles
var browseForever = function browseForever() {
    visits++;
    return cleanBrowser() //Destroy the zombie environment before each cycle
        .delay(5000).then(function(){
            global.gc();
            if (visits > maxVisits) {
                return Promise.delay(1000)
                    .then(function(){
                        global.gc();
                        console.log(process.memoryUsage());
                        return Promise.delay(1000);
                    });
            }

            initBrowser(); //Build a working zombie environment
            return browser.visit(testPageURL)
                .then(function(){return browser.wait();}) //Make sure we wait for everything to process (though the example demonstrates it's not good enough)
                .tap(function(){console.log('visit ' + visits + ': ' + process.memoryUsage().heapTotal);})
                .then(function(){return browseForever();});
        }).catch(function(e){
            console.error(e);
            return cleanBrowser()
                .then(function(){return browseForever();})
        });
}

//Manually wait for all the requests to process
var drainExternal = function drainExternal(){
    if (!cleanMemory) return Promise.resolve();

    var start = new Date();

    console.log('Draining pipeline - START');
    return Promise.all(pendingRequests.map(function(reqPromise){
      return reqPromise.reflect();
    })).tap(function(){console.log('Draining pipeline - END. Duration: ' + (new Date() - start)/1000);});
}

//Destroy zombie
var cleanBrowser = function cleanBrowser(){
    if (!browser) return Promise.resolve(); //Nothing to destroy

    //Do manual timer cleanup (clearTimers will do nothing if cleanMemory = false)
    for (var i=0; i<browser.tabs.length; i++) {
        clearTimers(browser.tabs[i]);
    }

    return browser.wait() //Make sure we only clean after zombie waits for the browse to complete
        .then(function(){return drainExternal();}) //Wait for pipeline to complete (drainExternal will do nothing if cleanMemory = false)
        .then(function(){
            //Clear additional timers that were created
            for (var i=0; i<browser.tabs.length; i++) {
                clearTimers(browser.tabs[i]);
            }

            return browser.wait(); //No one can blame us for not waiting enough
        })
        .then(function(){

            //Close each tab
            while (browser.tabs.length > 0){
                browser.tabs[0].close();
            }

            //Then destroy everything just in case
            browser.destroy();

            browser = null;
            Browser = null;

            global.gc(); //Force GC

            return Promise.resolve();
        });
};

//Manually clear timers
var clearTimers = function clearTimers(window){
    if (!cleanMemory) return;

    //Recursively run through sub-frames
    for (var i=0; i<window.length; i++){
        clearTimers(window[i]);
        window[i].document.close();
        window[i].close();
    }

    //Clear all pending timeouts, intervals and immediates
    var index;
    index = window.setTimeout(function(){}, 1000);
    while (index >= -1) {
        window.clearTimeout(index);
        index--;
    }
    index = window.setInterval(function(){}, 1000);
    while (index >= -1) {
        window.clearInterval(index);
        index--;
    }

    index = window.setImmediate(function(){}, 1000);
    while (index >= -1) {
        window.clearImmediate(index);
        index--;
    }

}

//Build the zombie environment
var initBrowser = function initBrowser(){

    Browser = require('zombie');
    browser = new Browser();

    browser.waitDuration='50s'; //The example site isn't known for its speed

    if (!cleanMemory) return; //All the following constructs are only needed for manual cleanups

    //Set up pipeline hooks for gathering requests and making sure they are completed
    pendingRequests = [];

    browser.pipeline.addHandler(function(b, req){

        if (req.pendingCompletion) req.pendingCompletion(); //This is a redirect. Resolve the only pending hook

        var myResolve;
        var myReject;
        var pending = new Promise(function(resolve, reject){myResolve=resolve; myReject = reject})
            .timeout(60000)
            .catch(function(e){
                myReject(e); console.error(e.toString() + ': ' + req.url)
            });

        req.pendingCompletion = myResolve; //Put resolver on the request for easy access from the response handler

        pendingRequests.push(pending); //Record pending hook
        return null;
    });

    browser.pipeline.addHandler(function(b, req, res){
        req.pendingCompletion(); //Resolve pending hook
        return res;
    });

}

browseForever(); //Initiate process
```

## The results

**Without** the manual cleanup, the heap prints out as:
visit 1: 141181952
visit 2: 174095616
visit 3: 221432320
visit 4: 253326080
visit 5: 296559104
visit 6: 340800000
visit 7: 378885376
visit 8: 491336704
visit 9: 419058688
visit 10: 466383360
{ rss: 480423936, heapTotal: 424060928, heapUsed: 344359288 }

**With** the manual cleanup (the pipeline drain duration indicates that indeed not all requests were processed):
visit 1: 131918592
Draining pipeline - START
Draining pipeline - END. Duration: 0.002
visit 2: 176147456
Draining pipeline - START
Draining pipeline - END. Duration: 0.156
visit 3: 173075712
Draining pipeline - START
Draining pipeline - END. Duration: 0.001
visit 4: 164844288
Draining pipeline - START
Draining pipeline - END. Duration: 2.011
visit 5: 177191424
Draining pipeline - START
Draining pipeline - END. Duration: 0.065
visit 6: 193757184
Draining pipeline - START
Draining pipeline - END. Duration: 0.002
visit 7: 163800320
Draining pipeline - START
Draining pipeline - END. Duration: 0.001
visit 8: 149401344
Draining pipeline - START
Draining pipeline - END. Duration: 0.001
visit 9: 184378880
Draining pipeline - START
Draining pipeline - END. Duration: 0.001
visit 10: 149401344
Draining pipeline - START
Draining pipeline - END. Duration: 0
{ rss: 103264256, heapTotal: 81364736, heapUsed: 45265880 }
